### DISCUSSION TITLE
Custom multi constructors

### DISCUSSION THREAD

**Discussion by kalemas** [redhat-developer/vscode-yaml#577] [General]
Hello, is it possible to handle yaml multi constructors? Now i got this 
![image](https://user-images.githubusercontent.com/638393/129321642-da37acfb-1108-4587-8d7d-a5503fa2a16c.png)
It is initialized with `PyYAML.Loader.add_multi_constructor()` and can handle any passed scalar, sequence or mapping.

**Comment by evidolob** [comment]
@kalemas Can you point to yaml multi constructors spec? As I can't find it definition in [YAML spec](https://yaml.org/spec/1.2/spec.html)

**Reply by kalemas** [reply]
In terms of YAML, python multi constructors should be implementation of tag resolution by common prefix (sorry, i may be wrong in details) that takes non fixed tag name and YAML node and return python object. YAML node would be any type.
Tag name would be dynamic in YAML afaik, so supporting wildcard in configuration would be useful. 
As an example i would give built in python object support that resolves to non fixed tag name:
```python
[23]: print(yaml.dump(type('Example', (), {}), canonical=True))
---
!!python/name:__main__.Example ""
```

**Reply by evidolob** [reply]
As I understand that is python specific, unfortunately we cannot support any features which not part of [YAML spec](https://yaml.org/spec/1.2/spec.html)

If you need to have custom tags, you can add them with `yaml.customTags` vscode preference.

**Reply by kalemas** [reply]
Yes, but i can not
1.  provide wildcard in tag name there (e.g. `!!python.name:*` as for case above)
2. bind custom tag to multiple types

I do not agree that this is python specific requests, as other implementations would use dynamic tags as well, i can find cases, if needed. And i do not see restrictions on node type followed be custom tag.

**Reply by evidolob** [reply]
But YAML spec doesn't contains dynamic tag notion.
So any support would be custom, and resolve problem for one, specific python library, but what if we will asked to add support for other library/language?

Can you force your library to include tag definition in to generated yaml?

**Reply by kalemas** [reply]
[Following](https://yaml.org/spec/1.2/spec.html#id2783273) may be best related to this case . So YAML tag is not something fixed but may be resolved by custom implementation (with still failsafe/fallback) to different object types depending on suffix.

This is not python specific feature, https://yaml.org/YAML_for_ruby.html look at `Struct class`. Yes native objects in yaml would not be desterilized, but there is fallback mechanism and custom tags would be resolved by multiple native implementations.

> Can you force your library to include tag definition in to generated yaml?

What do you mean? Can i get PyYAML implementation to produce yaml prefixed tags? Yes the above example produce prefix tags so if you change object type you will get different tag name:
```python
>>> import yaml
>>> yaml.dump(object())
'!!python/object:builtins.object {}\n'
>>> yaml.dump(frozenset())
'!!python/object/apply:builtins.frozenset\n- []\n'
```

**Reply by kalemas** [reply]
Or may be you need fully qualified tag name, with uri? I will make that.

**Reply by evidolob** [reply]
>What do you mean?

I was thinking in wrong direction, ignore that question.


Currently we cannot add support if wildcard in tags, as tag validation is on yaml parser, and it can take only predefined list of custom tags. The only possible solution is adding all custom tags in vscode preferences.

**Reply by kalemas** [reply]
So that just in default namespace
https://github.com/yaml/pyyaml/blob/ee37f4653c08fc07aecff69cfd92848e6b1a540e/lib3/yaml/constructor.py#L734

**Reply by kalemas** [reply]
No problem, i was just about to raising that case. I hope it will be ever possible to use tag prefixes but now i have to disable validation for YAML in vs code.
I will try to setup whole set of functions (its about 10-15 for my current project).
Thank you.

**Reply by kalemas** [reply]
including all used prefxed tags in settings resolved validation issues for custom tags... Supporting prefix tags is still nice to have.

**Reply by kalemas** [reply]
![image](https://user-images.githubusercontent.com/638393/129600120-cff1f06c-f852-4513-8d75-9777bd9beac0.png)

**Comment by Mathieson** [comment]
It would be nice to have a way to suppress/skip tag validation rather than disable the validation feature entirely. I'm running into the same thing right now with multiconstructors, and adding every tag variation to this list isn't a great option.
